Multiple retainedComponent (RootComponent) ComponentContext for android app will crash

Here my **RootComponent** class 

```
class RootComponent(componentContext: ComponentContext) : ComponentContext by componentContext {
    private val navigation = StackNavigation<Configuration>()

    val childStack = childStack(
        key = "mainRoot",
        source = navigation,
        serializer = Configuration.serializer(),
        initialConfiguration = Configuration.SplashActivity,
        handleBackButton = true,
        childFactory = ::createRootChild
    )

    private fun createRootChild(
        configuration: Configuration, context: ComponentContext
    ): Child {
        return when (configuration) {
            is Configuration.SplashActivity -> Child.SplashActivity(
                SplashComponent(contextComponent = context, onTimerComplete = {
                    navigation.pushNew(Configuration.LoginActivity)
                })
            )

            is Configuration.LoginActivity -> Child.LoginActivity(
                LoginComponent(componentContext = context, onLogin = {
                    navigation.pushNew(Configuration.DashboardActivity)
                })
            )

            is Configuration.DashboardActivity -> Child.DashboardActivity(DashboardComponent(context))
        }
    }

    sealed class Child {
        data class SplashActivity(val component: SplashComponent) : Child()
        data class LoginActivity(val component: LoginComponent) : Child()
        data class DashboardActivity(val component: DashboardComponent) : Child()
    }

    @Serializable
    sealed class Configuration {
        @Serializable
        data object SplashActivity : Configuration()

        @Serializable
        data object LoginActivity : Configuration()

        @Serializable
        data object DashboardActivity : Configuration()
    }
}
```


Here my **DashBoardRootComponent** class 

```
class DashBoardRootComponent(componentContext: ComponentContext) :
    ComponentContext by componentContext {
    private val navigation = StackNavigation<Configuration>()

    val childStack = childStack(
        key = "dashBoardRoot",
        source = navigation,
        serializer = Configuration.serializer(),
        initialConfiguration = Configuration.PlayListActivity,
        handleBackButton = true,
        childFactory = ::createDashBoardChild
    )

    private fun createDashBoardChild(
        configuration: Configuration, context: ComponentContext
    ): Child {
        return when (configuration) {
            is Configuration.PlayListActivity -> Child.PlayListActivity(
                PlayListComponent(componentContext = context, onPlayListClick = {
                    navigation.pushNew(Configuration.TrackListActivity)
                })
            )

            is Configuration.TrackListActivity -> Child.TrackListActivity(TrackListComponent(componentContext = context,
                onTrackListClick = {
                },
                onBackPress = {
                    navigation.pop()
                }))
        }
    }

    sealed class Child {
        data class PlayListActivity(val component: PlayListComponent) : Child()
        data class TrackListActivity(val component: TrackListComponent) : Child()
//        data class DashboardActivity(val component: DashboardComponent) : Child()
    }

    @kotlinx.serialization.Serializable
    sealed class Configuration {
        @kotlinx.serialization.Serializable
        data object PlayListActivity : Configuration()

        @kotlinx.serialization.Serializable
        data object TrackListActivity : Configuration()
    }

}
```

My App Component

```
@Composable
@Preview
fun App(root: RootComponent, dashBoardRoot: DashBoardRootComponent) {
    MaterialTheme {
        val childStack by root.childStack.subscribeAsState()
        Children(
            stack = childStack,
            animation = stackAnimation(slide()),
            content = { child ->
                when (val instance = child.instance) {
                    is RootComponent.Child.SplashActivity -> Splash(instance.component)
                    is RootComponent.Child.LoginActivity -> Login(instance.component)
                    is RootComponent.Child.DashboardActivity -> Dashboard(
                        instance.component,
                        dashBoardRootComponent = dashBoardRoot
                    )
                }
            }
        )
    }
}
```

For **Android MainActivity 1st Try** 

```
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val rootF = retainedComponent {
            RootComponent(it)
        }
        val dashF = retainedComponent {
            DashBoardRootComponent(it)
        }
        setContent {
            App(rootF, dashF)
        }
    }
}
```

Logcat  : 
```
FATAL EXCEPTION: main
Process: com.neonmusic.app, PID: 3633
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.neonmusic.app/com.neonmusic.app.MainActivity}: java.lang.IllegalArgumentException: SavedStateProvider with the given key is already registered
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7839)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.IllegalArgumentException: SavedStateProvider with the given key is already registered
	at androidx.savedstate.SavedStateRegistry.registerSavedStateProvider(SavedStateRegistry.kt:110)
	at com.arkivanov.essenty.statekeeper.AndroidExtKt.StateKeeper(AndroidExt.kt:31)
	at com.arkivanov.essenty.statekeeper.AndroidExtKt.stateKeeper(AndroidExt.kt:54)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent(RetainedComponent.kt:101)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent(RetainedComponent.kt:52)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent$default(RetainedComponent.kt:45)
	at com.neonmusic.app.MainActivity.onCreate(MainActivity.kt:20)
	at android.app.Activity.performCreate(Activity.java:8051)
	at android.app.Activity.performCreate(Activity.java:8031)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
	at android.os.Handler.dispatchMessage(Handler.java:106) 
	at android.os.Looper.loopOnce(Looper.java:201) 
	at android.os.Looper.loop(Looper.java:288) 
	at android.app.ActivityThread.main(ActivityThread.java:7839) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)  
```


For **Android MainActivity 2nd Try** from chatGpt reference

```
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val root = retainedComponentWithKey(key = "RootComponent_mainRoot") {
            RootComponent(it)
        }
        val dashBoardRoot = retainedComponentWithKey(key = "DashBoardRootComponent_dashBoardRoot") {
            DashBoardRootComponent(it)
        }

        setContent {
            App(root.first, dashBoardRoot.first)
        }
    }

    private fun <T : Any> retainedComponentWithKey(
        key: String,
        factory: (ComponentContext) -> T
    ): Pair<T, String> {
        val component = retainedComponent(key = key, handleBackButton = true, factory = factory)
        return component to key
    }
}
```

Logcat  : 
```
FATAL EXCEPTION: main
Process: com.neonmusic.app, PID: 3248
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.neonmusic.app/com.neonmusic.app.MainActivity}: java.lang.IllegalArgumentException: SavedStateProvider with the given key is already registered
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7839)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.IllegalArgumentException: SavedStateProvider with the given key is already registered
	at androidx.savedstate.SavedStateRegistry.registerSavedStateProvider(SavedStateRegistry.kt:110)
	at com.arkivanov.essenty.statekeeper.AndroidExtKt.StateKeeper(AndroidExt.kt:31)
	at com.arkivanov.essenty.statekeeper.AndroidExtKt.stateKeeper(AndroidExt.kt:54)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent(RetainedComponent.kt:101)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent(RetainedComponent.kt:52)
	at com.arkivanov.decompose.RetainedComponentKt.retainedComponent$default(RetainedComponent.kt:45)
	at com.neonmusic.app.MainActivity.retainedComponentWithKey(MainActivity.kt:33)
	at com.neonmusic.app.MainActivity.onCreate(MainActivity.kt:20)
	at android.app.Activity.performCreate(Activity.java:8051)
	at android.app.Activity.performCreate(Activity.java:8031)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
	at android.os.Handler.dispatchMessage(Handler.java:106) 
	at android.os.Looper.loopOnce(Looper.java:201) 
	at android.os.Looper.loop(Looper.java:288) 
	at android.app.ActivityThread.main(ActivityThread.java:7839) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)  
```


I am facing this issue only for android 

Do code for desktop also but it will working fine.
Here Desktop entry point

```
fun main() = application {
    Window(
        onCloseRequest = ::exitApplication,
        title = "neon-music-kmp",
    ) {
        val root = remember { RootComponent(DefaultComponentContext(LifecycleRegistry())) }
        val dashBoardRoot = remember { DashBoardRootComponent(DefaultComponentContext(LifecycleRegistry())) }
        App(root, dashBoardRoot)
    }
}
```

Can any one explain me where i need to give unique key for android retainedComponent ?
Tell me if i am wrong in anything.

--- a/decompose/src/androidMain/kotlin/com/arkivanov/decompose/RetainedComponent.kt
+++ b/decompose/src/androidMain/kotlin/com/arkivanov/decompose/RetainedComponent.kt
@@ -1,5 +1,6 @@
 package com.arkivanov.decompose
 
+import android.annotation.SuppressLint
 import androidx.activity.BackEventCompat
 import androidx.activity.ComponentActivity
 import androidx.activity.OnBackPressedCallback
@@ -97,13 +98,23 @@ internal fun <T, O> O.retainedComponent(
     isChangingConfigurations: () -> Boolean,
     factory: (ComponentContext) -> T,
 ): T where O : LifecycleOwner, O : SavedStateRegistryOwner, O : ViewModelStoreOwner {
-    val lifecycle = essentyLifecycle()
-    val stateKeeper = stateKeeper(discardSavedState = discardSavedState, isSavingAllowed = isStateSavingAllowed)
-    val marker = stateKeeper.consume(key = KEY_STATE_MARKER, strategy = String.serializer())
-    stateKeeper.register(key = KEY_STATE_MARKER, strategy = String.serializer()) { "marker" }
-    val instanceKeeper = instanceKeeper(discardRetainedInstances = marker == null)
+    check(savedStateRegistry.getSavedStateProvider(key = key) == null) {
+        "Another retained component is already registered with the key: $key"
+    }
 
-    check(!stateKeeper.isRegistered(key = key)) { "Another retained component is already registered with the key: $key" }
+    val lifecycle = essentyLifecycle()
+    val stateKeeper = stateKeeper(key = key, discardSavedState = discardSavedState, isSavingAllowed = isStateSavingAllowed)
+    val instanceKeeper = instanceKeeper()
+
+    val discardRetainedInstance = stateKeeper.consume(key = KEY_STATE_MARKER, strategy = Boolean.serializer()) == null
+    stateKeeper.register(key = KEY_STATE_MARKER, strategy = Boolean.serializer()) { true }
+
+    if (discardRetainedInstance) {
+        (instanceKeeper.remove(key = key) as RetainedComponentHolder<*>?)?.also { holder ->
+            holder.lifecycle.destroy()
+            holder.onDestroy()
+        }
+    }
 
     val holder =
         instanceKeeper.getOrCreate(key = key) {
@@ -154,14 +165,17 @@ private class DelegateOnBackPressedCallback(
         dispatcher.onBackPressed()
     }
 
+    @SuppressLint("VisibleForTests")
     override fun handleOnBackStarted(backEvent: BackEventCompat) {
         dispatcher.dispatchOnBackStarted(backEvent)
     }
 
+    @SuppressLint("VisibleForTests")
     override fun handleOnBackProgressed(backEvent: BackEventCompat) {
         dispatcher.dispatchOnBackProgressed(backEvent)
     }
 
+    @SuppressLint("VisibleForTests")
     override fun handleOnBackCancelled() {
         dispatcher.dispatchOnBackCancelled()
     }

--- a/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/RetainedComponentMultipleTest.kt
+++ b/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/RetainedComponentMultipleTest.kt
@@ -1,0 +1,334 @@
+package com.arkivanov.decompose
+
+import com.arkivanov.decompose.router.TestInstance
+import com.arkivanov.essenty.backhandler.BackCallback
+import com.arkivanov.essenty.instancekeeper.getOrCreate
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import kotlin.test.assertFalse
+import kotlin.test.assertNotSame
+import kotlin.test.assertNull
+import kotlin.test.assertSame
+import kotlin.test.assertTrue
+
+@Suppress("TestFunctionName")
+@RunWith(RobolectricTestRunner::class)
+class RetainedComponentMultipleTest {
+
+    @Test
+    fun WHEN_same_key_reused_THEN_throws_IllegalStateException() {
+        val owner = TestOwner()
+        owner.retainedComponent(key = "key")
+
+        assertFailsWith<IllegalStateException> {
+            owner.retainedComponent(key = "key")
+        }
+    }
+
+    @Test
+    fun WHEN_created_THEN_lifecycles_resumed() {
+        val owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        val events1 = ctx1.lifecycle.logEvents()
+        val events2 = ctx2.lifecycle.logEvents()
+
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events1)
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events2)
+    }
+
+    @Test
+    fun WHEN_recreated_THEN_old_lifecycle_destroyed() {
+        var owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        val events1 = ctx1.lifecycle.logEvents()
+        events1.clear()
+        val events2 = ctx2.lifecycle.logEvents()
+        events2.clear()
+
+        owner = owner.recreate(isChangingConfigurations = false)
+        owner.retainedComponent(key = "key1")
+        owner.retainedComponent(key = "key2")
+
+        assertContentEquals(listOf("onPause", "onStop", "onDestroy"), events1)
+        assertContentEquals(listOf("onPause", "onStop", "onDestroy"), events2)
+    }
+
+    @Test
+    fun WHEN_recreated_THEN_new_lifecycle_resumed() {
+        var owner = TestOwner()
+        owner.retainedComponent(key = "key1")
+        owner.retainedComponent(key = "key2")
+
+        owner = owner.recreate(isChangingConfigurations = false)
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val events1 = ctx1.lifecycle.logEvents()
+        val ctx2 = owner.retainedComponent(key = "key2")
+        val events2 = ctx2.lifecycle.logEvents()
+
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events1)
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events2)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_THEN_lifecycle_not_called() {
+        var owner = TestOwner()
+        val ctx = owner.retainedComponent()
+        val events = ctx.lifecycle.logEvents()
+        events.clear()
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        owner.retainedComponent()
+
+        assertContentEquals(emptyList(), events)
+    }
+
+    @Test
+    fun WHEN_recreated_THEN_saves_and_restores_states() {
+        var owner = TestOwner()
+        var ctx1 = owner.retainedComponent(key = "key1")
+        var ctx2 = owner.retainedComponent(key = "key2")
+        ctx1.stateKeeper.register(key = "key") { "saved_state_1" }
+        ctx2.stateKeeper.register(key = "key") { "saved_state_2" }
+
+        owner = owner.recreate(isChangingConfigurations = false)
+        ctx1 = owner.retainedComponent(key = "key1")
+        ctx2 = owner.retainedComponent(key = "key2")
+        val restoredState1 = ctx1.stateKeeper.consume<String>(key = "key")
+        val restoredState2 = ctx2.stateKeeper.consume<String>(key = "key")
+
+        assertEquals("saved_state_1", restoredState1)
+        assertEquals("saved_state_2", restoredState2)
+    }
+
+    @Test
+    fun retains_instances() {
+        var owner = TestOwner()
+        val ctx11 = owner.retainedComponent(key = "key1")
+        val ctx12 = owner.retainedComponent(key = "key2")
+
+        owner = owner.recreate()
+        val ctx21 = owner.retainedComponent(key = "key1")
+        val ctx22 = owner.retainedComponent(key = "key2")
+
+        assertSame(ctx11, ctx21)
+        assertSame(ctx12, ctx22)
+    }
+
+    @Test
+    fun GIVEN_isStateSavingAllowed_is_false_on_save_WHEN_recreated_THEN_state_not_saved() {
+        var owner = TestOwner()
+        var ctx1 = owner.retainedComponent(key = "key1", isStateSavingAllowed = { false })
+        var ctx2 = owner.retainedComponent(key = "key2", isStateSavingAllowed = { false })
+        ctx1.stateKeeper.register(key = "key") { "saved_state_1" }
+        ctx2.stateKeeper.register(key = "key") { "saved_state_2" }
+
+        owner = owner.recreate(isChangingConfigurations = false)
+        ctx1 = owner.retainedComponent(key = "key1")
+        ctx2 = owner.retainedComponent(key = "key2")
+        val restoredState1 = ctx1.stateKeeper.consume<String>(key = "key")
+        val restoredState2 = ctx2.stateKeeper.consume<String>(key = "key")
+
+        assertNull(restoredState1)
+        assertNull(restoredState2)
+    }
+
+    @Test
+    fun GIVEN_isStateSavingAllowed_is_false_on_save_WHEN_configuration_changed_THEN_instances_not_retained() {
+        var owner = TestOwner()
+        var ctx1 = owner.retainedComponent(key = "key1", isStateSavingAllowed = { false })
+        var ctx2 = owner.retainedComponent(key = "key2", isStateSavingAllowed = { false })
+        val instance11 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance12 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        ctx1 = owner.retainedComponent(key = "key1")
+        ctx2 = owner.retainedComponent(key = "key2")
+        val instance21 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance22 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        assertNotSame(instance11, instance21)
+        assertNotSame(instance12, instance22)
+    }
+
+    @Test
+    fun GIVEN_isStateSavingAllowed_is_false_on_save_WHEN_configuration_changed_THEN_old_instances_destroyed() {
+        var owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1", isStateSavingAllowed = { false })
+        val ctx2 = owner.retainedComponent(key = "key2", isStateSavingAllowed = { false })
+        val instance1 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance2 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        owner.retainedComponent(key = "key1")
+        owner.retainedComponent(key = "key2")
+
+        assertTrue(instance1.isDestroyed)
+        assertTrue(instance2.isDestroyed)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_discards_saved_state() {
+        var owner = TestOwner()
+        var ctx1 = owner.retainedComponent(key = "key1")
+        var ctx2 = owner.retainedComponent(key = "key2")
+        ctx1.stateKeeper.register(key = "key") { "saved_state_1" }
+        ctx2.stateKeeper.register(key = "key") { "saved_state_2" }
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        ctx1 = owner.retainedComponent(key = "key1", discardSavedState = true)
+        ctx2 = owner.retainedComponent(key = "key2", discardSavedState = true)
+        val restoredState1 = ctx1.stateKeeper.consume<String>(key = "key")
+        val restoredState2 = ctx2.stateKeeper.consume<String>(key = "key")
+
+        assertNull(restoredState1)
+        assertNull(restoredState2)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_instances_not_retained() {
+        var owner = TestOwner()
+        var ctx1 = owner.retainedComponent(key = "key1")
+        var ctx2 = owner.retainedComponent(key = "key2")
+        val instance11 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance12 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        ctx1 = owner.retainedComponent(key = "key1", discardSavedState = true)
+        ctx2 = owner.retainedComponent(key = "key2", discardSavedState = true)
+        val instance21 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance22 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        assertNotSame(instance11, instance21)
+        assertNotSame(instance12, instance22)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_old_instances_destroyed() {
+        var owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        val instance1 = ctx1.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+        val instance2 = ctx2.instanceKeeper.getOrCreate(key = "key", factory = ::TestInstance)
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        owner.retainedComponent(key = "key1", discardSavedState = true)
+        owner.retainedComponent(key = "key2", discardSavedState = true)
+
+        assertTrue(instance1.isDestroyed)
+        assertTrue(instance2.isDestroyed)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_old_instance_lifecycle_destroyed() {
+        var owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        val events1 = ctx1.lifecycle.logEvents()
+        events1.clear()
+        val events2 = ctx2.lifecycle.logEvents()
+        events2.clear()
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        owner.retainedComponent(key = "key1", discardSavedState = true)
+        owner.retainedComponent(key = "key2", discardSavedState = true)
+
+        assertContentEquals(listOf("onPause", "onStop", "onDestroy"), events1)
+        assertContentEquals(listOf("onPause", "onStop", "onDestroy"), events2)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_new_instance_lifecycle_resumed() {
+        var owner = TestOwner()
+        owner.retainedComponent()
+
+        owner = owner.recreate(isChangingConfigurations = true)
+        val ctx = owner.retainedComponent(discardSavedState = true)
+        val events = ctx.lifecycle.logEvents()
+
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events)
+    }
+
+    @Test
+    fun GIVEN_enabled_BackCallback_registered_in_all_retained_components_WHEN_onBackPressed_THEN_last_callback_called() {
+        val owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        var isCalled1 = false
+        var isCalled2 = false
+        ctx1.backHandler.register(BackCallback { isCalled1 = true })
+        ctx2.backHandler.register(BackCallback { isCalled2 = true })
+
+        owner.onBackPressedDispatcher.onBackPressed()
+
+        assertFalse(isCalled1)
+        assertTrue(isCalled2)
+    }
+
+    @Test
+    fun GIVEN_enabled_BackCallback_registered_in_first_retained_component_WHEN_onBackPressed_THEN_first_callback_called() {
+        val owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        var isCalled1 = false
+        var isCalled2 = false
+        ctx1.backHandler.register(BackCallback { isCalled1 = true })
+        ctx2.backHandler.register(BackCallback(isEnabled = false) { isCalled2 = true })
+
+        owner.onBackPressedDispatcher.onBackPressed()
+
+        assertTrue(isCalled1)
+        assertFalse(isCalled2)
+    }
+
+    @Test
+    fun GIVEN_enabled_BackCallback_registered_in_second_retained_component_WHEN_onBackPressed_THEN_second_callback_called() {
+        val owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        var isCalled1 = false
+        var isCalled2 = false
+        ctx1.backHandler.register(BackCallback(isEnabled = false) { isCalled1 = true })
+        ctx2.backHandler.register(BackCallback(isEnabled = true) { isCalled2 = true })
+
+        owner.onBackPressedDispatcher.onBackPressed()
+
+        assertFalse(isCalled1)
+        assertTrue(isCalled2)
+    }
+
+    @Test
+    fun GIVEN_disabled_BackCallback_registered_in_all_retained_components_WHEN_onBackPressed_THEN_callbacks_not_called() {
+        val owner = TestOwner()
+        val ctx1 = owner.retainedComponent(key = "key1")
+        val ctx2 = owner.retainedComponent(key = "key2")
+        var isCalled1 = false
+        var isCalled2 = false
+        ctx1.backHandler.register(BackCallback(isEnabled = false) { isCalled1 = true })
+        ctx2.backHandler.register(BackCallback(isEnabled = false) { isCalled2 = true })
+
+        owner.onBackPressedDispatcher.onBackPressed()
+
+        assertFalse(isCalled1)
+        assertFalse(isCalled2)
+    }
+
+    private fun TestOwner.retainedComponent(
+        key: String = "key",
+        discardSavedState: Boolean = false,
+        isStateSavingAllowed: () -> Boolean = { true },
+    ): ComponentContext =
+        retainedComponent(
+            key = key,
+            onBackPressedDispatcher = onBackPressedDispatcher,
+            discardSavedState = discardSavedState,
+            isStateSavingAllowed = isStateSavingAllowed,
+            isChangingConfigurations = { isChangingConfigurations },
+            factory = { it },
+        )
+}

--- a/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/RetainedComponentSingleTest.kt
+++ b/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/RetainedComponentSingleTest.kt
@@ -16,7 +16,7 @@ import kotlin.test.assertTrue
 
 @Suppress("TestFunctionName")
 @RunWith(RobolectricTestRunner::class)
-class RetainedComponentTest {
+class RetainedComponentSingleTest {
 
     @Test
     fun WHEN_created_THEN_lifecycle_resumed() {
@@ -55,7 +55,7 @@ class RetainedComponentTest {
     @Test
     fun WHEN_configuration_changed_THEN_lifecycle_not_called() {
         var owner = TestOwner()
-        val ctx = owner.retainedComponent(isChangingConfigurations = { true })
+        val ctx = owner.retainedComponent()
         val events = ctx.lifecycle.logEvents()
         events.clear()
 
@@ -166,6 +166,33 @@ class RetainedComponentTest {
     }
 
     @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_old_instance_lifecycle_destroyed() {
+        var owner = TestOwner()
+        val ctx = owner.retainedComponent()
+        val events = ctx.lifecycle.logEvents()
+        events.clear()
+
+        owner.isChangingConfigurations = true
+        owner = owner.recreate()
+        owner.retainedComponent(discardSavedState = true)
+
+        assertContentEquals(listOf("onPause", "onStop", "onDestroy"), events)
+    }
+
+    @Test
+    fun WHEN_configuration_changed_and_discardSavedState_is_true_on_restore_THEN_new_instance_lifecycle_resumed() {
+        var owner = TestOwner()
+        owner.retainedComponent()
+
+        owner.isChangingConfigurations = true
+        owner = owner.recreate()
+        val ctx = owner.retainedComponent(discardSavedState = true)
+        val events = ctx.lifecycle.logEvents()
+
+        assertContentEquals(listOf("onCreate", "onStart", "onResume"), events)
+    }
+
+    @Test
     fun GIVEN_enabled_BackCallback_registered_WHEN_onBackPressed_THEN_callback_called() {
         val owner = TestOwner()
         val ctx = owner.retainedComponent()
@@ -180,7 +207,7 @@ class RetainedComponentTest {
     @Test
     fun GIVEN_disabled_BackCallback_registered_WHEN_onBackPressed_THEN_callback_not_called() {
         val owner = TestOwner()
-        val ctx = owner.defaultComponentContext()
+        val ctx = owner.retainedComponent()
         var isCalled = false
         ctx.backHandler.register(BackCallback(isEnabled = false) { isCalled = true })
 
@@ -192,14 +219,13 @@ class RetainedComponentTest {
     private fun TestOwner.retainedComponent(
         discardSavedState: Boolean = false,
         isStateSavingAllowed: () -> Boolean = { true },
-        isChangingConfigurations: () -> Boolean = { false },
     ): ComponentContext =
         retainedComponent(
             key = "key",
             onBackPressedDispatcher = onBackPressedDispatcher,
             discardSavedState = discardSavedState,
             isStateSavingAllowed = isStateSavingAllowed,
-            isChangingConfigurations = isChangingConfigurations,
+            isChangingConfigurations = { isChangingConfigurations },
             factory = { it },
         )
 }

--- a/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/TestOwner.kt
+++ b/decompose/src/androidUnitTest/kotlin/com/arkivanov/decompose/TestOwner.kt
@@ -22,12 +22,15 @@ class TestOwner(
     override val savedStateRegistry: SavedStateRegistry get() = savedStateRegistryController.savedStateRegistry
     override val onBackPressedDispatcher: OnBackPressedDispatcher = OnBackPressedDispatcher()
 
+    var isChangingConfigurations: Boolean = true
+
     init {
         savedStateRegistryController.performRestore(savedState)
         lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
     }
 
     fun recreate(isChangingConfigurations: Boolean = true): TestOwner {
+        this.isChangingConfigurations = isChangingConfigurations
         val bundle = Bundle()
         savedStateRegistryController.performSave(bundle)
         val savedState = bundle.parcelize().deparcelize()

--- a/deps.versions.toml
+++ b/deps.versions.toml
@@ -2,7 +2,7 @@
 
 decompose = "3.2.2"
 kotlin = "2.0.20"
-essenty = "2.2.1"
+essenty = "2.3.0"
 reaktive = "2.2.0"
 junit = "4.13.2"
 jetbrainsCompose = "1.7.0"

